qan/analyzer/mongo/profiler/aggregator/aggregator.go

```diff
@@ -49,51 +52,169 @@ type Aggregator struct {
+	running      bool            // Is this service running?
+	doneChan     chan struct{}   // close(doneChan) to notify goroutines that they should shutdown
+	wg           *sync.WaitGroup // Wait() for goroutines to stop after being notified they should shutdown
 }
 
-// Add aggregates new system.profile document and returns report if it's ready
-func (self *Aggregator) Add(doc proto.SystemProfile) (*qan.Report, error) {
+// Add aggregates new system.profile document
+func (self *Aggregator) Add(doc proto.SystemProfile) error {
 	self.Lock()
 	defer self.Unlock()
+	if !self.running {
+		return fmt.Errorf("aggregator is not running")
+	}
 
 	ts := doc.Ts.UTC()
 
 	// skip old metrics
 	if ts.Before(self.timeStart) {
-		return nil, nil
+		self.stats.DocsSkippedOld.Add(1)
+		return nil
 	}
 
-	return self.interval(ts), self.stats.Add(doc)
+	// if new doc is outside of interval then finish old interval and flush it
+	if !ts.Before(self.timeEnd) {
+		self.flush(ts)
+	}
+
+	// we had some activity so reset timer
+	self.t.Reset(self.d)
+
+	// add new doc to stats
+	self.stats.DocsIn.Add(1)
+	return self.mongostats.Add(doc)
 }
 
-// Report generates report for current interval and starts new one
-func (self *Aggregator) Report() *qan.Report {
+func (self *Aggregator) Start() <-chan *qan.Report {
 	self.Lock()
 	defer self.Unlock()
+	if self.running {
+		return self.reportChan
+	}
 
-	return self.interval(time.Now())
+	// create new channels over which we will communicate to...
+	// ... outside world by sending collected docs
+	self.reportChan = make(chan *qan.Report, ReportChanBuffer)
+	// ... inside goroutine to close it
+	self.doneChan = make(chan struct{})
+
+	// set status
+	self.stats = &stats{}
+	self.status = status.New(self.stats)
+
+	// timeout after not receiving data for interval time
+	self.t = time.NewTimer(self.d)
+
+	// start a goroutine and Add() it to WaitGroup
+	// so we could later Wait() for it to finish
+	self.wg = &sync.WaitGroup{}
+	self.wg.Add(1)
+	go start(
+		self.wg,
+		self,
+		self.doneChan,
+		self.stats,
+	)
+
+	self.running = true
+	return self.reportChan
 }
 
-// interval sets interval if necessary and returns *qan.Report for old interval if not empty
-func (self *Aggregator) interval(ts time.Time) *qan.Report {
-	// if time is before interval end then we are still in the same interval, nothing to do
-	if ts.Before(self.timeEnd) {
+func (self *Aggregator) Stop() {
+	self.Lock()
+	defer self.Unlock()
+	if !self.running {
+		return
+	}
+	self.running = false
+
+	// notify goroutine to close
+	close(self.doneChan)
+
+	// wait for goroutines to exit
+	self.wg.Wait()
+
+	// close reportChan
+	close(self.reportChan)
+}
+
+func (self *Aggregator) Status() map[string]string {
+	self.RLock()
+	defer self.RUnlock()
+	if !self.running {
 		return nil
 	}
 
+	return self.status.Map()
+}
+
+func start(
+	wg *sync.WaitGroup,
+	aggregator *Aggregator,
+	doneChan <-chan struct{},
+	stats *stats,
+) {
```

Comment:
The single-line version of it will be short enough (shorted that the comments below).